File: frontend/src/types/activity.ts
/**
 * Types for the activity timeline surface (PR #58).
 *
 * Mirrors `backend/app/schemas/activity.py`. Both endpoints
 * (`/api/requests/{id}/activity` and `/api/contracts/{id}/activity`)
 * return the same item shape.
 *
 * The endpoint is read-only; rows project from the existing audit log
 * and never mutate state.
 */

export type ActivityEventType =
  | "approval.workflow.created"
  | "approval.step.activated"
  | "approval.step.approved"
  | "approval.step.rejected"
  | "approval.workflow.completed"
  | "approval.workflow.rejected"
  | "approval.workflow.cancelled"
  | "contract.sent_for_signature"
  | "contract.executed"
  | string;

export interface ActivityTimelineItem {
  id: string;
  event_type: ActivityEventType;
  /** ISO timestamp. */
  occurred_at: string;
  actor_user_id: string | null;
  /** Server-rendered label; clients should not derive this. */
  title: string;
  /** Optional second-line description (server-rendered). */
  description: string | null;
  request_id: string | null;
  contract_id: string | null;
  workflow_run_id: string | null;
  approval_step_id: string | null;
  step_order: number | null;
  /** "ad_hoc" / "template" / "policy" — workflow origin label. */
  source: string | null;
}

export interface ActivityTimelineResponse {
  items: ActivityTimelineItem[];
}

File: backend/app/schemas/activity.py
"""Schemas for activity timeline endpoints (PR #58).

Read-only chronological feed for a request or a contract. Items are
projected from ``AuditEvent`` rows — every nested model uses
``extra="forbid"`` and only allowlists the safe identifier set, so
``storage_key`` / ``wrapped_dek`` / signer PII / DocuSeal secrets
cannot leak even if a future audit detail accidentally carried one.

The timeline is **explainability only**: the endpoint never mutates
state, never auto-creates workflows, never changes the gate's allow/
block rules.
"""
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ActivityTimelineItem(BaseModel):
    """One row in the activity feed.

    ``title`` is a short human-readable label rendered server-side so
    every client renders the same string (no client-side i18n / format
    drift). ``description`` is an optional second line. ``metadata`` is
    a small allowlisted projection of the underlying audit event's
    ``details`` — never the raw ``details`` blob.
    """

    model_config = ConfigDict(extra="forbid")

    id: uuid.UUID
    event_type: str
    occurred_at: datetime
    actor_user_id: uuid.UUID | None = None
    title: str
    description: str | None = None
    request_id: uuid.UUID | None = None
    contract_id: uuid.UUID | None = None
    workflow_run_id: uuid.UUID | None = None
    approval_step_id: uuid.UUID | None = None
    step_order: int | None = None
    source: str | None = None


class ActivityTimelineResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    items: list[ActivityTimelineItem] = Field(default_factory=list)
